working/macros/feature-specification.md

```diff
@@ -113,66 +113,116 @@ visible. For example, if two macros applied to two methods in the same class,
 application order can be detected.
 
-## Implementation Overview - Multi-Phase Approach
+### Introspection of macro modifications
 
-At a high level the idea is to build the program up in a series of steps with
-macro phases interleaved between them. At each phase, macros only have access
-to the parts of the program that are already complete, and can produce
-information for later phases. Later phases generally get more introspective
-power, but less power to mutate the program.
+Imagine you have the following:
 
-In general, the introspection APIs are limited such that only things produced
-by previous phases can be introspected upon. This ensures we always give
-accurate and complete information in these APIs.
+```dart
+@serialize class Foo {
+  Bar bar;
+}
+
+class Bar {
+  @memoize int three() => 3;
+}
+```
 
-A macro can contribute code in multiple phases if needed. A common use case for
-this is running in an early phase to define the signature of a new declaration,
-and then filling in the implementation of that declaration in a later phase when
-more introspection power is available.
+The `@memoize` macro on Bar adds a field, `_memo`, to the surrounding class to
+store the memoized value. The `@serialize` on Foo generates a `toJson()` method
+that recursively serializes all of the fields on Foo. Since it does so
+recursively, it would end up traversing into and introspecting on the fields of
+Bar. When that happens, does the `@serialize` macro see the `_memo` field added
+by `@memoize`? The answer depends on which macro runs first. But, since these
+are unrelated macros on essentially unrelated declarations, that order shouldn't
+be user visible.
 
-When a macro adds a new declaration, it may also add a macro application to that
-declaration, but only if that macro runs in a *later* phase than the current
-one.
+Our solution to this is stratify macro application into phases. The
+introspection API is restricted so that macros within a phase cannot see any
+changes made by other macros in the same phase. The phases are described in
+detail below.
 
-### Phase 1: Type Macros
+### Complete macro application order
 
-In this phase, macros are allowed to contribute entirely new types to the
-program. These could come in the form of classes, typedefs, enums, etc.
+When all of these are put together, an idealized compilation and macro
+application of a Dart program looks like this:
 
-Since new types can be contributed to the program in this phase, very little
-reflective power is provided to these macros. You can see the names of types
-that are referenced in the declaration for instance, but you can't ask if they
-are a subtype of a known type, because we don't know how the types will resolve
-yet. Even a type which could be resolved to an existing type might not actually
-resolve to that type once macros are done (a new type could be introduced which
-shadows the original one).
+1.  For each library, ordered topologically by imports:
```

Comment:
Is it intended that all the itens are `1.`?

Reply:
github at least renders these with actual sequential numbers when you do it like this (it makes writing the markdown a bit easier this way).

Reply:
Yeah, Markdown automatically renumbers them.

Reply:
Oh, I didn't know that.